### PR TITLE
Check if volume supports exchangedata before use

### PR DESCRIPTION
**Pull request by antons**
This fixes incompatibility with APFS. As suggested in #320, I extracted a C function. If you’d like to see anything else changed, let me know.

I AGREE TO THE GITUP CONTRIBUTOR LICENSE AGREEMENT.

### CODE REVIEW

File: GitUpKit/Core/GCFunctions.h [hunk 1]
```diff
@@ -39,6 +39,8 @@ NSString* GCFileSystemPathFromGitPath(const char* string);
 NSURL* GCURLFromGitURL(NSString* url);
 NSString* GCGitURLFromURL(NSURL* url);
 
+int CGExchangeDataFromTempPathToPath(const char* tempPath, const char* path);
```
Comment (swisspol) [inline GitUpKit/Core/GCFunctions.h]: Should be called `GCExchangeFileData(const char* path1, const char* path2)`: this function is unrelated to temp files.

File: GitUpKit/Core/GCFunctions.m [hunk 2]
```diff
@@ -131,6 +135,27 @@ BOOL GCGitOIDFromSHA1Prefix(NSString* prefix, git_oid* oid, NSError** error) {
   return url.absoluteString;
 }
 
+int CGExchangeDataFromTempPathToPath(const char* tempPath, const char* path) {
+  int status;
+  struct statfs stat;
+  status = statfs(path, &stat);
```
Comment (swisspol) [inline GitUpKit/Core/GCFunctions.m]: Move `int status` to this line

Why are you calling `statfs` here anyway?

File: GitUpKit/Core/GCFunctions.m [hunk 3]
```diff
@@ -131,6 +135,27 @@ BOOL GCGitOIDFromSHA1Prefix(NSString* prefix, git_oid* oid, NSError** error) {
...
+    status = getattrlist(stat.f_mntonname, &attrList, &attrBuf, sizeof(attrBuf), 0);
+    if (status == 0) {
+      if ((attrBuf.attr.capabilities[VOL_CAPABILITIES_INTERFACES] & VOL_CAP_INT_EXCHANGEDATA) == VOL_CAP_INT_EXCHANGEDATA) {
+        status = exchangedata(tempPath, path, FSOPT_NOFOLLOW);
+      } else {
+        status = rename(tempPath, path);
+      }
+    }
+  } else {
+    status = rename(tempPath, path);
```
Comment (swisspol) [inline GitUpKit/Core/GCFunctions.m]: Remove this statement: if `statfs` fails, then the function should fail immediately as this would be a fatal error

File: GitUpKit/Core/GCLiveRepository.m [hunk 4]
```diff
@@ -500,15 +500,8 @@ - (void)_writeSnapshots {
...
-      struct stat info;
-      if (lstat(path.fileSystemRepresentation, &info) == 0) {
-        if (exchangedata(tempPath.fileSystemRepresentation, path.fileSystemRepresentation, FSOPT_NOFOLLOW) == 0) {
-          success = YES;
-        }
-      } else {
-        if (rename(tempPath.fileSystemRepresentation, path.fileSystemRepresentation) == 0) {
-          success = YES;
-        }
+      if (CGExchangeDataFromTempPathToPath(tempPath.fileSystemRepresentation, path.fileSystemRepresentation) == 0) {
```
Comment (swisspol) [inline GitUpKit/Core/GCLiveRepository.m]: You changed the logic of the code here.

**Comment by antons** [comment]
@swisspol `statfs` is used to get `stat.f_mntonname` for `getattrlist`. 

Updates you requested are on the branch.

Didn’t realize you wanted CLA signed with every pull request. (I already signed it with a previously accepted pull request.)

File: GitUpKit/Core/GCFunctions.m [hunk 5]
```diff
@@ -131,6 +135,25 @@ BOOL GCGitOIDFromSHA1Prefix(NSString* prefix, git_oid* oid, NSError** error) {
   return url.absoluteString;
 }
 
+int GCExchangeFileData(const char* path1, const char* path2) {
```
Comment (antons) [inline GitUpKit/Core/GCFunctions.m]: This doesn’t seem to be the right to name for a function that can rename the file at `path1` to `path2`. Original name at least suggested that the file is supposed to be temporary.
